vertical-pod-autoscaler/pkg/recommender/util/binary_decaying_histogram.go

```diff
@@ -0,0 +1,275 @@
+	}
+	if h.retentionDays != o.retentionDays {
+		panic("can't merge histograms with different retention days")
+	}
+	for day := o.lastDayIndex; day > o.lastDayIndex-o.retentionDays && day >= 0; day-- {
+		bucket := o.bucketForDay[day%o.retentionDays]
+		if day <= h.lastDayIndex && day > h.lastDayIndex-h.retentionDays {
+			if h.bucketForDay[day%h.retentionDays] > bucket {
+				bucket = h.bucketForDay[day%h.retentionDays]
+			}
+		}
+		if bucket > 0 {
+			h.addSampleToBucket(bucket, day)
+		}
+	}
+}
+
+func (h *binaryDecayingHistogram) Percentile(percentile float64) float64 {
+	if percentile != 1.0 {
+		panic("Only Percentile(1.0) function is implemented for binary decaying histogram, got: " + fmt.Sprintf("%f", percentile))
+	}
+	maxBucket := uint16(0)
+	for day := 0; day < h.retentionDays; day++ {
+		if h.bucketForDay[day] > maxBucket {
+			maxBucket = h.bucketForDay[day]
+		}
+	}
+	value := 0.0
+	if maxBucket > 0 {
+		value = h.valueForBucket(int(maxBucket) - 1)
+	}
+	return value
+}
+
+func (h *binaryDecayingHistogram) IsEmpty() bool {
+	for day := 0; day < h.retentionDays; day++ {
+		if h.bucketForDay[day] != 0 {
+			return false
+		}
+	}
+	return true
+}
+
+func (h *binaryDecayingHistogram) valueForBucket(bucket int) float64 {
+	if bucket < h.options.NumBuckets()-1 {
+		// Return the end of the bucket.
+		return h.options.GetBucketStart(bucket + 1)
+	}
+	// Return the start of the last bucket (note that the last bucket
+	// doesn't have an upper bound).
+	return h.options.GetBucketStart(bucket)
+}
+
+func (h *binaryDecayingHistogram) String() string {
+	lines := []string{
+		fmt.Sprintf("retentionDays: %d", h.retentionDays),
+		"day\tbucket\tvalue",
+	}
+	for day := h.lastDayIndex; day > h.lastDayIndex-h.retentionDays && day >= 0; day-- {
+		bucket := h.bucketForDay[day%h.retentionDays]
+		value := 0.0
+		if bucket > 0 {
+			value = h.valueForBucket(int(bucket) - 1)
+		}
+		lines = append(lines, fmt.Sprintf("%d\t%d\t%.3f", day, bucket, value))
+	}
+	lines = append(lines, "\n")
+	return strings.Join(lines, "\n")
+}
+
+func (h *binaryDecayingHistogram) Equals(other Histogram) bool {
+	h2, typesMatch := other.(*binaryDecayingHistogram)
+	if !typesMatch || h.options != h2.options || h.retentionDays != h2.retentionDays || h.lastDayIndex != h2.lastDayIndex {
+		return false
+	}
+	for day := 0; day < h.retentionDays; day++ {
+		if h.bucketForDay[day] != h2.bucketForDay[day] {
+			return false
+		}
+	}
+	return true
+}
+
+// Saving to checkpoint is done by converting in memory representation (list of days) to list of bucket weights.
+// One day requires 1 bit of storage, so we can store 32 days in one uint32. Since number of days can be arbitrary,
+// we may need multiple uint32s to store all days for one bucket. The bits for bucket 0 are stored in checkpoint with
+// index 0, h.options.NumBuckets(), 2*h.options.NumBuckets() etc. The bits for bucket 1 are stored in checkpoint with
+// index 1, h.options.NumBuckets()+1, 2*h.options.NumBuckets()+1 etc.
```

Comment:
Cool! So if I'm trying to store 40 days and there are 7 buckets, then the checkpoint will look like the following?
```
bucketWeights:
 0: <bitset storing which of the first 32 days are of bucket 0>
 1: <bitset storing which of the first 32 days are of bucket 1>
 ....
 7: <bitset storing which of the first 32 days are of bucket 7>
 8: <bitset storing which of the last 8 days are of bucket 0>
 ....
```

Reply:
Yes except that I believe you have off-by-one in bucket names, if there are 7 buckets and buckets are named starting from zero then the last bucket is `bucket 6` and example would look like this:
```
bucketWeights:
 0: <bitset storing which of the first 32 days are of bucket 0>
 1: <bitset storing which of the first 32 days are of bucket 1>
 ....
 6: <bitset storing which of the first 32 days are of bucket 6>
 7: <bitset storing which of the last 8 days are of bucket 0>
 ....
```